### ISSUE TITLE
Obligations not propagating in subgroups

### ISSUE THREAD

**Issue by copycat-killer** [moise-lang/moise#9]
In my JaCaMo project, there is a MAS starting from one agent, `alice`. This agent includes functionalities of another agent, named `executor`, which in turn creates the agent `bob` via the default internal action. Then, `bob` creates agent `carl` via a custom CArtAgO artifact. `alice` (via `executor`) starts an organisation with a main group and a subgroup. `alice` and `bob` join the main group, while `carl` joins the subgroup. `alice` creates a scheme and commits to a mission, which involves `bob` and `carl`. `carl` should have the first obligation in the mission, but it doesn't: no obligation belief appears in its BB.

This is `organisation.xml` file. Full project can be downloaded [here](https://github.com/moise-lang/moise/files/1185622/example.zip).

```xml
<?xml version="1.0" encoding="UTF-8"?>

<?xml-stylesheet href="http://moise.sourceforge.net/xml/os.xsl" type="text/xsl"?>

<organisational-specification
    id="organisation"
    os-version="0.8"

    xmlns='http://moise.sourceforge.net/os'
    xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
    xsi:schemaLocation='http://moise.sourceforge.net/os
                        http://moise.sourceforge.net/xml/os.xsd'>

    <structural-specification>
        <role-definitions>
            <role id="executor"/>
            <role id="bob"/>
            <role id="carl"/>
        </role-definitions>

        <group-specification id="mainGroup">
            <roles>
                <role id="executor"/>
                <role id="bob"/>
            </roles>
            <subgroups>
                <group-specification id="subGroup">
                    <roles>
                        <role id="carl"/>
                    </roles>
                </group-specification>
            </subgroups>
        </group-specification>
    </structural-specification>

    <functional-specification>
        <scheme id="doSomething">
            <goal id="do_executor">
                <plan operator="parallel">
                    <goal id="do_bob">
                        <plan operator="parallel">
                            <goal id="do_carl"/>
                        </plan>
                    </goal>
                </plan>
            </goal>
            <mission id="mExecutor">
                <goal id="do_executor"/>
            </mission>
            <mission id="mBob">
                <goal id="do_bob"/>
            </mission>
            <mission id="mCarl">
                <goal id="do_carl"/>
            </mission>
        </scheme>
    </functional-specification>

    <normative-specification>
        <norm type="permission" role="executor" mission="mExecutor" id="n1"/>
        <norm type="obligation" role="bob"      mission="mBob"      id="n2"/>
        <norm type="obligation" role="carl"     mission="mCarl"     id="n3"/>
    </normative-specification>

</organisational-specification>
```

This is a screenshot of the scheme board debug. As you can see, `do_carl` plan (which belongs to `carl`) is enabled, but `carl` does not commit to its role, thus the scheme is stuck.

![example](https://user-images.githubusercontent.com/1777696/28752634-da481e3c-7524-11e7-8651-b979f201f3c7.png)

What am I doing wrong? Maybe I should extend the responsibility of `myscheme` to subgroup? But how?

Thanks in advance.

---

While we are at it, I have a practical question: is there a way to reuse schemes? That is, create only once, then reset and execute again arbitrary times? For now, I remove and recreate, but it seems computationally expensive.

**Comment by jomifred** [comment]
Hi, is it possible for you to package your code and send it to me? (jomifred@gmail.com) The context is quite complex to figure out the problem without more details. 

Regarding reusing schemes, you can use the operation "resetGoal(<put the root goal here>)". It also works to "backtrack" the execution of some scheme to some point.

**Comment by copycat-killer** [comment]
> Hi, is it possible for you to package your code and send it to me?

Done.

**Comment by jomifred** [comment]
Hi, thanks for the code. There are some minor issues:

1. we have to link the two groups to define one as sub-group of another. In bob code:

	createGroup(subgroup, subGroup, GrArtId); 
	setParentGroup(maingroup)[artifact_id(GrArtId)]; // **** added
	….


2. in carl.asl, the common code for moise should be included so that carl focus on the normative board of the organisation (the one the obliges him to commit to the mission)

	{ include("$jacamoJar/templates/common-moise.asl") } // **** added


3. there is a bug in create agent (already fixed in the sources of github and soon in the jacamo release). To fix, in executor.asl

	jacamo.create_agent("bob", "bob.asl").  // instead of .create_agent

and in BobEnv.java

	rs.createAgent("carl", "carl.asl", null, new ArrayList<String>(Arrays.asList("jaca.CAgentArch", "jacamo.infra.JaCaMoAgArch")), null, null, null); rs.startAgent("carl”);


4. to oblige an agent to commit to a mission, its cardinality should be greater than 0, so in org.xml (otherwise, it is just an permission)

	<mission id="mCarl" min="1”> 

I hope I listed all changes required to run the application.

Jomi

**Comment by jomifred** [comment]
Regarding the bug with create agent, I've just replaced the jacamo distribution file with the fix at https://sourceforge.net/projects/jacamo/files/version-0/jacamo-0.7-SNAPSHOT.zip/download

**Comment by copycat-killer** [comment]
Thanks Jomi, everything is working now. 

I attach the fixed example for the benefit of future readers: [example_fixed.zip](https://github.com/moise-lang/moise/files/1185957/example_fixed.zip)

Points 1 and 4 seemed counterintuitive at first, but thinking about it I then realized their reason.

Have a nice sunday evening!

**Comment by copycat-killer** [comment]
@jomifred 

Sorry to bother you again, but I have another issue related to this organisation.

I have a new test case, you can download it here: [example.zip](https://github.com/moise-lang/moise/files/1195084/example.zip)

In this test case, I modified the MAS so that `executor` creates 2 instance of `bob`, named `bob1` and `bob2`. Each `bob` then creates its own `carl`.

But when I start the MAS, tipically the obligations are not all respected. An output example:

```
CArtAgO Http Server running on http://127.0.0.1:3273
Jason Http Server running on http://127.0.0.1:3272
Moise Http Server running on http://127.0.0.1:3271
[bob2_carl] I am obliged to commit to mCarl on myscheme... doing so
[bob2] I am obliged to commit to mBob on myscheme... doing so
[bob1] I am obliged to commit to mBob on myscheme... doing so
[bob1] I am obliged to commit to mBob on myscheme... doing so
[alice] done
```
Here, `bob1_carl` obligation message is missing, while `bob1` prints twice.

I have to mention that there are times when the messages are right, that is something like this:

```
CArtAgO Http Server running on http://127.0.0.1:3273
Jason Http Server running on http://127.0.0.1:3272
Moise Http Server running on http://127.0.0.1:3271
[bob1_carl] I am obliged to commit to mCarl on myscheme... doing so
[bob2_carl] I am obliged to commit to mCarl on myscheme... doing so
[bob1] I am obliged to commit to mBob on myscheme... doing so
[bob2] I am obliged to commit to mBob on myscheme... doing so
[alice] done
```

But these times are very rare.

What am I doing wrong now? Thanks in advance as always.

**Comment by jomifred** [comment]
The cardinality of mission mCarl is 1, it means that the scheme requires 1 mCarl. Once the first Carl commits to it, the scheme is fine and no more agents (the second Carl) are obliged to commit to it.

You can either: 
1. change the cardinality to 2
2. code the agent Carl to explicitly commit to the mission

**Comment by copycat-killer** [comment]
Isn't there a way to specify the cardinality of a mission to be at least 1?

Also, can mission cardinalities be set from a scheme property? For instance, I'd like to add a scheme X with N mCarl missions, where N is defined at runtime.

**Comment by jomifred** [comment]
Hi, some agent can do the action 

	admCommand(“setCardinality(mission,themissionId,min,max)”);

to dynamically change the cardinality of some mission. The argument is a string, that you will likely use the internal action .concat to built.

**Comment by copycat-killer** [comment]
Thanks for your answers. Closing definitely.
